<b><center>IMP LINKS</center></b>

https://keras.io/keras_tuner/

https://www.tensorflow.org/tutorials/keras/keras_tuner

https://github.com/tensorflow/docs/blob/master/site/en/tutorials/keras/keras_tuner.ipynb

https://keras.io/getting_started/

-------

In [2]:
import pandas as pd
import keras_tuner
from tensorflow import keras
from tensorflow.keras import layers
from keras_tuner.tuners import RandomSearch
from keras.layers import Dense 
from keras.layers import LeakyReLU,ReLU,ELU
from keras.layers import Dropout

In [3]:
dt=pd.read_csv("https://raw.githubusercontent.com/krishnaik06/Keras-Tuner/main/Real_Combine.csv")

In [4]:
dt=dt.dropna()

In [5]:
X= dt.iloc[:,:-1]
y= dt.iloc[:,-1]

In [6]:
X

T    TM    Tm     SLP     H   VV     V    VM
0      7.4   9.8   4.8  1017.6  93.0  0.5   4.3   9.4
1      7.8  12.7   4.4  1018.5  87.0  0.6   4.4  11.1
2      6.7  13.4   2.4  1019.4  82.0  0.6   4.8  11.1
3      8.6  15.5   3.3  1018.7  72.0  0.8   8.1  20.6
4     12.4  20.9   4.4  1017.3  61.0  1.3   8.7  22.2
...    ...   ...   ...     ...   ...  ...   ...   ...
1088  18.1  24.0  11.2  1015.4  56.0  1.8  15.9  25.9
1089  17.8  25.0  10.7  1015.8  54.0  2.3   9.4  22.2
1090  13.9  24.5  11.4  1015.0  95.0  0.6   8.7  14.8
1091  16.3  23.0   9.8  1016.9  78.0  1.1   7.4  16.5
1092  16.3  23.4   9.0  1017.3  68.0  1.3   7.8  18.3

[1092 rows x 8 columns]

In [7]:
y

0       219.720833
1       182.187500
2       154.037500
3       223.208333
4       200.645833
           ...    
1088    288.416667
1089    256.833333
1090    169.000000
1091    186.041667
1092    185.583333
Name: PM 2.5, Length: 1092, dtype: float64

# Hyperparameters
- How many number of hidden layers we should have?
- How many number of neurons we should have in hidden layers?
- Learning Rate

In [8]:
def build_model(hp):
    model = keras.Sequential()
    for i in range(hp.Int('num_layers', 2, 20)):
        model.add(layers.Dense(units=hp.Int('units_' + str(i),
                                            min_value=32,
                                            max_value=512,
                                            step=32),
                               activation='relu'))
    model.add(layers.Dense(1, activation='linear'))
    model.compile(
        optimizer=keras.optimizers.Adam(
            hp.Choice('learning_rate', [1e-2, 1e-3, 1e-4])),
        loss='mean_absolute_error',
        metrics=['mean_absolute_error'])
    return model

In [9]:
tuner = RandomSearch(
    build_model,
    objective=keras_tuner.Objective("val_mean_absolute_error", direction="min"),
    max_trials=5,
    executions_per_trial=3,
    directory='project',
    project_name='Air Quality Index')

INFO:tensorflow:Reloading Oracle from existing project project\Air Quality Index\oracle.json
INFO:tensorflow:Reloading Tuner from project\Air Quality Index\tuner0.json


In [10]:
tuner.search_space_summary()

Search space summary
Default search space size: 20
num_layers (Int)
{'default': None, 'conditions': [], 'min_value': 2, 'max_value': 20, 'step': 1, 'sampling': None}
units_0 (Int)
{'default': None, 'conditions': [], 'min_value': 32, 'max_value': 512, 'step': 32, 'sampling': None}
units_1 (Int)
{'default': None, 'conditions': [], 'min_value': 32, 'max_value': 512, 'step': 32, 'sampling': None}
learning_rate (Choice)
{'default': 0.01, 'conditions': [], 'values': [0.01, 0.001, 0.0001], 'ordered': True}
units_2 (Int)
{'default': None, 'conditions': [], 'min_value': 32, 'max_value': 512, 'step': 32, 'sampling': None}
units_3 (Int)
{'default': None, 'conditions': [], 'min_value': 32, 'max_value': 512, 'step': 32, 'sampling': None}
units_4 (Int)
{'default': None, 'conditions': [], 'min_value': 32, 'max_value': 512, 'step': 32, 'sampling': None}
units_5 (Int)
{'default': None, 'conditions': [], 'min_value': 32, 'max_value': 512, 'step': 32, 'sampling': None}
units_6 (Int)
{'default': None, 'co

In [11]:
from sklearn.model_selection import train_test_split
Xtrain,Xtest,ytrain,ytest=train_test_split(X,y,test_size=.3,random_state=0)

In [12]:
tuner.search(Xtrain,ytrain,
            epochs=5,
            validation_data=(Xtest,ytest))

INFO:tensorflow:Oracle triggered exit


In [13]:
tuner.search_space_summary()

Search space summary
Default search space size: 20
num_layers (Int)
{'default': None, 'conditions': [], 'min_value': 2, 'max_value': 20, 'step': 1, 'sampling': None}
units_0 (Int)
{'default': None, 'conditions': [], 'min_value': 32, 'max_value': 512, 'step': 32, 'sampling': None}
units_1 (Int)
{'default': None, 'conditions': [], 'min_value': 32, 'max_value': 512, 'step': 32, 'sampling': None}
learning_rate (Choice)
{'default': 0.01, 'conditions': [], 'values': [0.01, 0.001, 0.0001], 'ordered': True}
units_2 (Int)
{'default': None, 'conditions': [], 'min_value': 32, 'max_value': 512, 'step': 32, 'sampling': None}
units_3 (Int)
{'default': None, 'conditions': [], 'min_value': 32, 'max_value': 512, 'step': 32, 'sampling': None}
units_4 (Int)
{'default': None, 'conditions': [], 'min_value': 32, 'max_value': 512, 'step': 32, 'sampling': None}
units_5 (Int)
{'default': None, 'conditions': [], 'min_value': 32, 'max_value': 512, 'step': 32, 'sampling': None}
units_6 (Int)
{'default': None, 'co

In [14]:
tuner.results_summary()

Results summary
Results in project\Air Quality Index
Showing 10 best trials
Trial summary
Hyperparameters:
num_layers: 17
units_0: 288
units_1: 384
learning_rate: 0.01
units_2: 32
units_3: 32
units_4: 32
units_5: 32
units_6: 32
units_7: 32
units_8: 32
units_9: 32
units_10: 32
units_11: 32
units_12: 32
units_13: 32
units_14: 32
units_15: 32
units_16: 32
Score: 56.602638244628906
Trial summary
Hyperparameters:
num_layers: 18
units_0: 128
units_1: 128
learning_rate: 0.001
units_2: 256
units_3: 256
units_4: 384
units_5: 352
units_6: 96
units_7: 320
units_8: 448
units_9: 224
units_10: 160
units_11: 128
units_12: 288
units_13: 256
units_14: 256
units_15: 320
units_16: 96
units_17: 32
Score: 61.177252451578774
Trial summary
Hyperparameters:
num_layers: 4
units_0: 288
units_1: 320
learning_rate: 0.0001
units_2: 288
units_3: 384
units_4: 512
units_5: 160
units_6: 480
units_7: 384
units_8: 192
units_9: 192
units_10: 352
units_11: 352
units_12: 384
units_13: 384
units_14: 192
units_15: 384
units_

In [15]:
best=tuner.get_best_models(num_models=1)[0]

Hyperparameters:   
num_layers: 17  
units_0: 288  
units_1: 384  
learning_rate: 0.01  
units_2: 32  
units_3: 32  
units_4: 32  
units_5: 32  
units_6: 32  
units_7: 32  
units_8: 32  
units_9: 32  
units_10: 32  
units_11: 32  
units_12: 32  
units_13: 32  
units_14: 32  
units_15: 32  
units_16: 32  
Score: 56.602638244628906  

In [16]:
best.summary

<bound method Model.summary of <keras.engine.sequential.Sequential object at 0x000001988855EEE0>>